### PR TITLE
Adds UseSanitizedBuildSystemEnvironment=YES

### PR DESCRIPTION
**Pull request by flovilmart**
And allows override with Process info

proposed fix for #992

### CODE REVIEW

File: Source/CarthageKit/Xcode.swift [hunk 1]
```diff
@@ -172,7 +172,12 @@ public struct BuildArguments {
 		// Disable code signing requirement for all builds
 		// Frameworks get signed in the copy-frameworks action
 		args += [ "CODE_SIGNING_REQUIRED=NO", "CODE_SIGN_IDENTITY=" ]
-
+		
+		// If we're in xcode phase, sanitize
+		let environment = NSProcessInfo.processInfo().environment
+		if let xcodeVersionActual = environment["XCODE_VERSION_ACTUAL"] as? String {
```
Comment (flovilmart) [inline Source/CarthageKit/Xcode.swift]: I used an environment file that is pretty Xcode specific, should not be set in any other environment. Suggestions welcome

**Comment by ikesyo** [comment]
Compiler complains:

```
/Users/travis/build/Carthage/Carthage/Source/CarthageKit/Xcode.swift:178:67: error: downcast from 'String?' to 'String' only unwraps optionals; did you mean to use '!'?
                if let xcodeVersionActual = environment["XCODE_VERSION_ACTUAL"] as? String {
```

File: Source/CarthageKit/Xcode.swift (hunk 1 above)
Comment (mdiep) [inline Source/CarthageKit/Xcode.swift]: As I commented in https://github.com/Carthage/Carthage/issues/992#issuecomment-172999671, I think we should set this argument all the time unless we've encountered an actual problem from setting it.

Comment (flovilmart) [inline Source/CarthageKit/Xcode.swift]: I from time to times use external arguments. Also, if we compare to other build systems/dependency manager (for ex npm), they all inherit from the current environment. I have mixed feelings about that for that reason.

Comment (mdiep) [inline Source/CarthageKit/Xcode.swift]: Would you mind sharing what things you use external arguments for?

Comment (flovilmart) [inline Source/CarthageKit/Xcode.swift]: Mostly with XC_CONFIG_FILE to override bitcode settings in order to have all the frameworks build with the same settings regardless what the developer wrote :)

File: Source/CarthageKit/Xcode.swift [hunk 2]
```diff
@@ -172,6 +172,10 @@ public struct BuildArguments {
 		// Disable code signing requirement for all builds
 		// Frameworks get signed in the copy-frameworks action
 		args += [ "CODE_SIGNING_REQUIRED=NO", "CODE_SIGN_IDENTITY=" ]
+		
+		// Allow to force unsanitized environment
+		let useSanitizedEnvironment = environment["XCODEBUILD_USE_SANITIZED_ENVIRONMENT"] ?? "YES"
```
Comment (ikesyo) [inline Source/CarthageKit/Xcode.swift]: Another complaint:

```
/Users/travis/build/Carthage/Carthage/Source/CarthageKit/Xcode.swift:177:33: error: use of unresolved identifier 'environment'
                let useSanitizedEnvironment = environment["XCODEBUILD_USE_SANITIZED_ENVIRONMENT"] ?? "YES"
```

Please build it on your local machine before pushing.

Comment (flovilmart) [inline Source/CarthageKit/Xcode.swift]: Sorry about that...

**Comment by hyperspacemark** [comment]
@mdiep Is there anything holding this up? Looks like #992 is the last thing preventing 0.12 from being released.

**Comment by mdiep** [comment]
This needs to be reviewed and tested. I haven't had time to do so yet.

**Comment by kballard** [comment]
LGTM, though it might be a bit odd that saying `env XCODEBUILD_USE_SANITIZED_ENVIRONMENT= carthage build` will suppress sanitizing (i.e. it would be nice to treat an empty string the same as having the env var being unset).

I haven't tested this in Carthage directly, but I did test the `-UseSanitizedBuildSystemEnvironment=YES` flag and it does behave as documented.

**Comment by flovilmart** [comment]
@mdiep you want support for XCODEBUILD_USE_SANITIZED_ENVIRONMENT= -> "NO" ?

**Comment by mdiep** [comment]
We should either always specify `-UseSanitizedBuildSystemEnvironment=YES` (which I'd lean towards) or add a command line flag to Carthage that defaults to using it. I don't think an environment variable is the right way to go for this.

**Comment by flovilmart** [comment]
I'd need a way to bypass the sanitized environment. I don't mind having and extra flag. That's actually pretty much just for myself XCODEBUILD_USE_SANITIZED_ENVIRONMENT=NO as no one seems to use custom configuration on the top of what carthage provides.

**Comment by mdiep** [comment]
Would you mind switching it to a boolean command line flag? `--(no-)sanitized-builds` or something?

**Comment by flovilmart** [comment]
@mdiep I definitely could, I'm not familiar with the command line parsing facility.

**Comment by mdiep** [comment]
It should be pretty easy. `BuildCommand.Options` lets you define a property. Then you'll need to add a flag to the build commands to pass in whether or not builds should be sanitized.

**Comment by kballard** [comment]
Seems to me that adding a flag to `BuildArguments` that has to be set is going to be tricky, because you have to propagate the flag all the way through the build process, and hope that you don't miss anything. You'd also want to sanitize the calls to `xcodebuild -list` and `xcodebuild -showBuildSettings` too for good measure (just in case the environment affects that), but the caller of `xcodebuild -list` at least doesn't have a good way of passing freeform arguments through, so e.g. `func schemesInProject(project: ProjectLocator)` would have to gain a new argument just to specify build sanitizing. And this all seems rather pointless, since what you actually want is to explicitly sanitize *all* calls to `xcodebuild` (or none of them) instead of controlling it on a call-by-call basis.

I'd suggest that you should use a static stored property on `BuildArguments` that can be set once after processing the command-line arguments, but AFAIK Carthage currently avoids any kind of global state, and if there's already multithreading going on before the command-line args have been parsed then it might not be a good idea to have global shared state (unless you add some form of synchronization around the property, but I don't know if Carthage has an easy way to do that at the moment).

**Comment by flovilmart** [comment]
@mdiep  the architecture for the build and archive command begs for a refactoring as they both share most of the same options, I'll have to duplicate the command line args options code, that seems pretty inefficient at the moment.

**Comment by kballard** [comment]
Given that this is blocking 0.12, which fixes an annoying iTunesConnect warning, I'd rather we just go ahead with the environment variable solution, and then consider doing a command-line flag later. Especially because very few people will want to not sanitize their build environment.

**Comment by mdiep** [comment]
I'm just going to take this out of the 0.12 milestone for now.

**Comment by flovilmart** [comment]
Agreed with @kballard, we should also add a common Options for the build and update before adding that flag to prevent even more code duplications.

File: Source/CarthageKit/Xcode.swift [hunk 3]
```diff
@@ -172,6 +172,11 @@ public struct BuildArguments {
 		// Disable code signing requirement for all builds
 		// Frameworks get signed in the copy-frameworks action
 		args += [ "CODE_SIGNING_REQUIRED=NO", "CODE_SIGN_IDENTITY=" ]
+		
+		// Allow to force unsanitized environment
```
Comment (NachoSoto) [inline Source/CarthageKit/Xcode.swift]: Could you expand this comment to explain what `XCODEBUILD_USE_SANITIZED_ENVIRONMENT` is exactly?

Comment (flovilmart) [inline Source/CarthageKit/Xcode.swift]: Starting Xcode 7, the UseSanitizedBuildEnvironment defaults to YES, which overrides/disables any externally set environment variables.

This env var allow a user to override the default behavior (=YES) in order to force xcodebuild to use the current environment. 

Is that clear enough? I've added that a while ago, and don't really recall what the status was on the issue/resolution we wanted to apply to the original issue.

Any guidance appreciated

**Comment by ikesyo** [comment]
Closing this due to inactivity. But feel free to succeed to this. We now have `BuildOptions` type so this should be easier than before.

**Pull request closed without merging**